walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,113 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
 
 public final class FindMeetingQuery {
+  /*
+  NOT_OCCUPIED < OPTIONAL_OCCUPIED < MANDATORY_OCCUPIED
+   */
+  private static final int NOT_OCCUPIED = 0;
+  private static final int OPTIONAL_OCCUPIED = 1;
+  private static final int MANDATORY_OCCUPIED = 2;
+
+  /*
+  Criterion functional interface is used to define who's (mandatory or all attendees')
+  occupied hours should be taken into account in the findAvailableTimeRanges function.
+   */
+  @FunctionalInterface
+  private interface Criterion {
+    boolean satisfies(int x);
+  }
+  private final Criterion COMMON_CRITERION = x -> x == NOT_OCCUPIED;
+  private final Criterion MANDATORY_CRITERION = x -> x == OPTIONAL_OCCUPIED || x == NOT_OCCUPIED;
+
+
+  private static final int MINUTES_IN_DAY = 60 * 24;
+
+  private int getStart(Event event) {
+    return event.getWhen().start();
+  }
+
+  private int getEnd(Event event) {
+    return event.getWhen().end();
+  }
+
+  private boolean haveCommonAttendees(Collection<String> a, Collection<String> b) {
+    return new HashSet<>(a).removeAll(b);
+  }
+
+  /**
+   * @param occupied shows what minutes are available for booking
+   *                 for mandatory (0|1) or all(0) attendees
+   * @param criterion defines whose (mandatory or all attendees')
+   *                  occupied hours should be taken in account
+   * @param duration the duration of the new meeting
+   * @return the list of time ranges available to schedule the meeting
+   *
+   *  The algorithm skips all minutes that don't meet the criterion,
```

Comment:
nit: put this section explaining the algorithm above the `@param` and `@return` section.